text_to_x/text_to_embeddings.py

```diff
@@ -0,0 +1,306 @@
+    
+    
+    ###
+    ### EMBEDDING OPERATIONS
+    ###
+    
+    def similarity(self, word1, word2):
+        '''
+        calcualte the dot product between two given embeddings. Order doesn't matter.
+        - word1: str, first word to query
+        - wrod2: str, second word to query
+        
+        Example: 
+        In: self.similarity('hpv', 'vaccine')
+        
+        Out: 0.11681473580251966
+        '''
+        
+        if not all(isinstance(i, str) for i in [word1, word2]):
+            raise ValueError(f"expected string, please input a string to query")
+        
+        if word1 in self.tok2ind and word2 in self.tok2ind:
+            index_word1 = self.tok2ind[word1]
+            index_word2 = self.tok2ind[word2]
+            return np.dot(self.model[index_word1], self.model[index_word2])
+        
+        else:
+            if word1 in self.tok2ind and word2 not in self.tok2ind:
+                raise ValueError(f"{word2}: query not in dataset")
+            elif word1 not in self.tok2ind and word2 in self.tok2ind:
+                raise ValueError(f"{word1}: query not in dataset")
+                
+                
+    def find(self, query):
+        '''
+        input word, get vector representation
+        
+        Example:
+        In: self.find('hpv')
+        
+        Out: array([ 0.00435369, ...,  0.332069  ])
+        '''
+        if query in self.tok2ind:
+            query_index = self.tok2ind[query]
+            return self.model[query_index]
+        
+        raise ValueError("the word you're querying is not present in the model")
+        
+                
+                
+    def similar_to_query(self, query, n_similar = 5):
+        '''
+        print n most similar words to query (default 5)
+        cosine similarity is the metric used (min = -1, max = 1)
+        
+        - query: str, word to query
+        - n_similar: int, number of most simialr words to output
+        
+        Example:
+        In: self.similar_to_query('hpv')
+        
+        Out: [(0.3425330655678359, 'type'),
+             (0.29492709699932357, 'eksempel'),
+             (0.2840084462156332, 'larynxcance'),
+             (0.2736081170128576, 'sygdom'),
+             (0.24797964395624894, 'forårsag')]
+        '''
+        
+        if not isinstance(query, str):
+            raise ValueError(f"query expected a string")
+            
+        if not isinstance(n_similar, int):
+            raise ValueError(f"n_similar expected int")
+            
+        
+        vec = self.find(query)
+        
+        # cosine similarity
+        cos_sim_matrix = np.dot(self.model, vec)/(np.linalg.norm(self.model)*np.linalg.norm(vec))
+        
+        list_similar = []
+
+        for i in np.argpartition(-1 * cos_sim_matrix, n_similar + 1)[:n_similar + 1]:
+
+            if self.ind2tok[i] == query:
+                continue
+
+            list_similar.append((float(cos_sim_matrix[i]), self.ind2tok[i]))
```

Comment:
Might be faster as list comprehension. E.g.:

```
list_similar = [(float(cos_sim_matrix[i]), self.ind2tok[i]) for i in \
                np.argpartition(-1 * cos_sim_matrix, n_similar + 1)[:n_similar + 1] \
                if self.ind2tok[i] != query]
```

Reply:
the \ are redundant though